content/blog/2020-06-29-optimizing-dvc-remotes.md

```diff
@@ -0,0 +1,193 @@
+picture: ''
+author: peter_rowlands
+---
+
+One of the key features provided by DVC is the ability to efficiently
+synchronize versioned datasets between a user's local machine and
+[remote storage](https://dvc.org/doc/command-reference/remote), and our users
+have frequently shown the need for us to optimize this process wherever
+possible.
+
+![](/uploads/images/2020-06-29/optimization_screenshot.png '=600')
+
+DVC 1.0 includes several performance improvements with regard to how DVC
+synchronizes data to and from remotes. In particular, for this release we
+focused on optimizing performance in situations where users with must regularly
+`dvc push` (relatively) small modifications made to large datasets versioned
+with DVC. This post provides an overview of some of the ways in which remote
+performance has been optimized in DVC 1.0.
+
+## Overview
+
+In order to sync data between a local machine and remote storage, we must first
+do the following (equivalent to the `dvc status` command with option `-c`
+(`--cloud`)), before data can actually be uploaded or downloaded to or from the
+remote:
+
+1. Determine which files are present locally (i.e. in the DVC cache).
+2. Determine which files are present in the remote.
+3. Determine the difference between the two sets of files.
+
+Commonly used cloud sync utilities, such as [rclone](https://rclone.org/), must
+be generalized to support any file structure, which can come at the cost of
+performance, especially when dealing with large datasets. In DVC, since we have
+control over the structure of both our local cache and our remote storage, we
+leverage this structure to optimize this "status" step in all remote storage
+operations (i.e. `dvc status -c`, `dvc push`, `dvc pull`, `dvc fetch`). In DVC
+version 1.0, these optimizations have significantly improved performance
+compared to older versions of DVC and tools like rclone.
+
+![benchmarks](https://raw.githubusercontent.com/gist/pmrowla/338d9645bd05df966f8aba8366cab308/raw/37f64e28e8c2e963aff0d320f6e0cea0724342a5/remote-benchmarks.svg)
+_Note: benchmark details can be found
+[here](https://gist.github.com/pmrowla/338d9645bd05df966f8aba8366cab308#benchmarks),
+rclone run using default configuration_
+
+## DVC cache/remote structure
+
+Before continuing, it will be helpful for the reader to keep in mind a few key
+points about the DVC cache and remote storage structure.
+
+```
+.dvc/cache
+├── 00
+│   ├── 411460f7c92d2124a67ea0f4cb5f85
+│   ├── 6f52e9102a8d3be2fe5614f42ba989
+│   └── ...
+├── 01
+├── 02
+├── 03
+├── ...
+└── ff
+```
+
+_Example DVC cache_
+
+- Files versioned by DVC are identified and stored according to their MD5 hash.
+  The cache is organized into 256 subdirectories (from `00` to `ff`), where each
+  subdirectory is the first 2 characters in an MD5 hash string.
+- MD5 is an evenly distributed hash function, so the DVC cache will be evenly
```

Comment:
Do we have a source that would back this statement? I think it would be great to link it if someone wants to get to nuts and bolts.